### PR TITLE
Sem-Ver: bugfix Reduce the time taken to generate a jwt by passing through a load private key instance to jwt.encode.

### PR DESCRIPTION
**Pull request by dbaxa**
Signed-off-by: David Black <dblack@atlassian.com>

### CODE REVIEW

**Comment by dbaxa** [comment]
We may consider this a (slight) api break.

File: atlassian_jwt_auth/signer.py [hunk 1]
```diff
@@ -50,7 +64,7 @@ def generate_jwt(self, audience, **kwargs):
             self.issuer)
         return jwt.encode(
             self._generate_claims(audience, **kwargs),
-            key=private_key_pem,
+            key=self._private_key,
```
Comment (dbaxa) [inline atlassian_jwt_auth/signer.py]: This almost halves the time taken to encode a RS256 jwt (after the first access to `_private_key`).

Comment (mark-adams) [reply atlassian_jwt_auth/signer.py]: Hmmm... I'm not getting the same results when I profile this function (master and your branch are essentially the same). I don't think the `@property` decorator does any sort of caching or memoization of the call; it is just syntatic sugar so you don't have to call the method with parenthesis.

I created a branch where we explicitly check for `self._private_key is None` and then populate it if it is not present. The benchmarks are about 50% lower with that change. I think that may have been what you were shooting for?

https://github.com/atlassian/asap-authentication-python/pull/92

**Comment by dbaxa** [comment]
cc @jbunton-atlassian - I think that the change that is being made here would arguably break `FilePrivateKeyRetriever` because we would load a private key only once. Instead this code would need to be changed to retain a cache based upon `key_identifier`.

**Comment by dbaxa** [comment]
I have updated the code to use a cache of private keys - to avoid the problem of having a private key retriever only be called once.

File: atlassian_jwt_auth/signer.py [hunk 2]
```diff
@@ -25,6 +28,25 @@ def __init__(self, issuer, private_key_retriever, **kwargs):
             raise ValueError("lifetime, '%s',exceeds the allowed 1 hour max" %
                              (self.lifetime))
 
+    def _obtain_private_key(self, key_identifier, private_key_pem):
```
Comment (mark-adams) [inline atlassian_jwt_auth/signer.py]: Nit: `obtain` is a little too fancy for me. Why not `get`?

Comment (dbaxa) [reply atlassian_jwt_auth/signer.py]: I prefer `obtain` over `get` in this case because it isn't a simple access.

**Comment by dbaxa** [comment]
Performance change -> 

Before this change
```
ec 3.3106096630217507
rsa 14.982443070970476
```

After this change
```
ec 2.271472151041962
rsa 8.497295178007334
```

Benchmarking script (based off @mark-adams 's found in https://github.com/atlassian/asap-authentication-python/pull/92) 
```
import timeit

import atlassian_jwt_auth
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric import rsa, ec
from cryptography.hazmat.primitives.serialization import Encoding
from cryptography.hazmat.primitives.serialization import PrivateFormat
from cryptography.hazmat.primitives import serialization


def wrapper(func, *args, **kwargs):
    def wrapped():
        return func(*args, **kwargs)
    return wrapped


def sign(signer):
    signer.generate_jwt('audience')


def main():
    ec_private_key = ec.generate_private_key(
        ec.SECP256R1(), default_backend())

    rsa_private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=2048,
        backend=default_backend()
    )

    ec_private_private_pem = ec_private_key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.TraditionalOpenSSL,
        encryption_algorithm=serialization.NoEncryption()
    )

    rsa_private_key_pem = rsa_private_key.private_bytes(
        Encoding.PEM, PrivateFormat.PKCS8, serialization.NoEncryption()
    )
    issuer = 'issuer'
    key_id = 'issuer/key'

    ec_signer = atlassian_jwt_auth.create_signer(
        issuer, key_id, ec_private_private_pem, algorithm='ES256'
    )
    rsa_signer = atlassian_jwt_auth.create_signer(
        issuer, key_id, rsa_private_key_pem, algorithm='RS256'
    )
    ec_w = wrapper(sign, ec_signer)
    rsa_w = wrapper(sign, rsa_signer)
    num = 10000
    print('ec', timeit.timeit(ec_w, number=num))
    print('rsa', timeit.timeit(rsa_w, number=num))


main()
```

**Pull request merged**